### ISSUE TITLE
Solaar tray-icon does not show + possible-fix

### ISSUE THREAD

**Issue by jwrdegoede** [pwr-Solaar/Solaar#1116]
While working on making the TopIcons GNOME extension, which is necessary to get tray-icons to work with GNOME3 run with GNOME40, I noticed that the solaar tray-icon only shows when running GNOME40 on Xorg and not when running GNOME40 as Wayland compositor.

This is caused by Solaar being GTK based and GTK using the Wayland backend when running under Wayland, which is all good, except that Wayland does not have anything like tray-icons, so the GTK tray-icon code is just stubs when running under Wayland.

I've tried and this can be worked around by forcing the use of the X11 GTK backend, and thus use Xwayland at which point the tray-icon compat stuff in GNOME40 happily picks up the tray-icon.

Specifically I tried killing the running solaar process and started it manually like this:
GDK_BACKEND=x11 solaar --window=hide

Forcing use of the x11 backend can be also done from within the code itself by calling gdk_set_allowed_backends() (or the PyGTK equivalent).

Note I'm not sure if this is the best solution, that depends on how much being native-wayland is valued over how much having the tray-icon is valued. One thing to keep in mind here is that at least under GNOME3 tray-icons won't be visible by default (they require installing an extension).

Alternatively an option would be to add some check to:
/etc/xdg/autostart/solaar.desktop

To have it not run when running under Wayland, because to me it seems that the primary purpose of having solaar autostart is showing the trayicon.

Maybe instead of --window=hide have a --tray-icon option for the /etc/xdg/autostart/solaar.desktop file and then check if the trayicon functionality is available and if it is not available then just exit. This would also work for when the TopIcon extension is not installed in GNOME3 independent of the windowing-system, because then the trayicon availability check in GTK will report that the trayicon functionality is not available.

**Comment by pfps** [comment]
Solaar will have to go through some changes for Wayland.   Not just for the tray but also for input emulation.  Solaar uses [Ayatana]AppIndicator3 as its interface to the tray and depends on it working correctly.  It appears that AppIndicator3 is not working correctly under Wayland, but what can Solaar do in this case?   Adding a special check for this bug doesn't appear to be a reasonable thing to do.

What is the preferred method for providing a small-screen-space interface to long-running utility programs in Gnome?

**Comment by pfps** [comment]
But then how does something like power-manager or a bluetooth manager run?  They need to be running so that they can provide notifications and perform certain operations in the background (for Solaar this includes setting up devices when they connect) but the user also needs to initiate operations (such as pairing new bluetooth devices).  How can the user do this if the program is completely invisible?

**Comment by jwrdegoede** [comment]
The daemon like functionality and the UI are split and the daemon has a dbus API to allow the UI to talk to it.

E.g. for bluetooth we have bluez + gnome-control-center code to allow pairing new devices.

And for power-management we have the upower daemon.

I realize that this may be overkill for solaar. Alternatively you may advice users to install the https://extensions.gnome.org/extension/3681/top-indicator-app/ GNOME extension this implements appindicator support, which I think should also work under Wayland. I don't use this extension myself; I guess the tray-icon stuff which I'm seeing with solaar is a fallback for when there is no appindicator support.

If you think advising solaar users to use this extension is a good workaround, then I can give it a try with solaar.

**Comment by pfps** [comment]
How then does the stuff on the top-right corner of the display work?  There are several icons there (network, sound, power).

Maybe, regrettably, this is only for things that are blessed by Gnome.  What happened to the idea of allowing users to decide what is important for them?

**Comment by pfps** [comment]
If you could try out the extension that would be good.  It could then be put into the Solaar documentation.

**Comment by jwrdegoede** [comment]
I've tested the gnome-shell-extension-appindicator extension and with that things indeed work fine under Wayland without any Solaar changes.

So this indeed is probably best fixed by documenting that Solaar needs the gnome-shell-extension-appindicator extension to be installed under GNOME3 for it to fully work. Note this is installed and enabled on Ubuntu by default.

For Fedora this can be installed by doing:

```
sudo dnf install gnome-shell-extension-appindicator
```

Followed by enabling the extension in the extensions-app which can be launched from the UI or by running "gnome-extensions-app" from a terminal.

**Comment by pfps** [comment]
Excellent.  Is this needed for all versions of Gnome or just those running with Wayland?

**Comment by jwrdegoede** [comment]
All versions of GNOME need either the appindicator extension (preferred modern way of dealing with this), or if that is not present then they need the TopIcons extenstion, as in that case libappindicator will fallback to using the old X11-only tray-icon stuff. But that only works under Wayland.

TL;DR: yes this is needed for all versions of GNOME3.

**Comment by pfps** [comment]
Do you know the command to install the extension in Ubuntu?

**Comment by pfps** [comment]
PR #1123 has the first cut at these instructions.   Does this seem reasonable to you?

**Comment by gitowiec** [comment]
Hi
I'm using Ubuntu Budgie 20.04 LTS. Solaar does not show up the app try indicator with just `solaar --window=hide` but when I do `GDK_BACKEND=x11 solaar --window=hide` it does. When I want to run it in background with `GDK_BACKEND=x11 solaar --window=hide &` the icon does not show up.

**Comment by pfps** [comment]
@jwrdegoede Please look at the documentation changes in PR #1123 and say whether they make sense to you.
